### PR TITLE
main:load csv data to database

### PR DESCRIPTION
**Pull request by Jiani-Li**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -1 +1,107 @@
...
+        conn = pyodbc.connect(
+        Trusted_Connection='Yes',
+        Driver='{ODBC Driver 17 for SQL Server}',
+        Server=server,
+        Database=database
+        )
+        cursor = conn.cursor()
+        return conn, cursor
+
+    except Exception as e:
```
Comment (Arman-Mojaver) [inline main.py]: It is better if you do not use a bare Exception here, either check how the try block can fail (and use that exception), or write a custom exception as follows:

class DatabaseConnectionError(Exception):
    pass

File: main.py [hunk 2]
```diff
@@ -1 +1,107 @@
...
+        print(f'Error connecting to the database: {e}')
+        return None, None
+
+def create_dbtable(original_file,cursor, conn):
```
Comment (Arman-Mojaver) [inline main.py]: create_database_table, or create_db_table.
fix the comma in original_file,

File: main.py [hunk 3]
```diff
@@ -1 +1,107 @@
...
+    try:
+        # Read the original csv data
+        masterlist = pd.read_csv(original_file, delimiter=';')
+        masterlist['Date'] = pd.to_datetime(masterlist['Date'])
+        #Create table in database
+        cursor.execute('''
```
Comment (Arman-Mojaver) [inline main.py]: Do not put the string directly, create a variable in the file, or a class variable if you are putting it in a class

File: main.py [hunk 4]
```diff
@@ -1 +1,107 @@
...
+
+    except Exception as e:
+        print(f'Error creating table: {e}')
+
+def load_original_data(original_file, cursor, conn):
+    try:
+        masterlist = pd.read_csv(original_file, delimiter=';')
+        for row in masterlist.itertuples():
+            cursor.execute(
+                '''
```
Comment (Arman-Mojaver) [inline main.py]: same here

File: main.py [hunk 5]
```diff
@@ -1 +1,107 @@
...
+                VALUES(?, ?, ?)
+                ''',
+                row.Date,
+                row.Server,
+                row.Cost
+            )
+        conn.commit()
+        print('Original Data is loaded.')
+
+    except Exception as e:
```
Comment (Arman-Mojaver) [inline main.py]: same here

File: main.py [hunk 6]
```diff
@@ -1 +1,107 @@
...
+# Calling the function to connect to the database
+conn, cursor = database_connect(server_name, database_name)
+# Calling the function to create table in database
+#create_dbtable(original_file, cursor, conn)
+# Calling the function to load the original data
+#load_original_data(original_file, cursor, conn)
+# Calling the function to load the incremental data
+load_incremental_data(new_file, cursor, conn)
+# Close the connection to the database
+conn.close()
```
Comment (Arman-Mojaver) [inline main.py]: dont execute the script directly, use `if __name__ == "__main__"`, and put it in a function

File: main.py [hunk 7]
```diff
@@ -1 +1,107 @@
...
+
+        # Filter out incremental data with dates greater than the maximum date in the database
+        new_data = incremental_list[incremental_list['Date'] > max_date]
+
+        # If there is new_data, insert the new_data into database
+        if not new_data.empty:
+            for row in new_data.itertuples():
+                cursor.execute(
+                    '''
+                    INSERT INTO [Python-SQL-CSV].[dbo].[masterlist]
```
Comment (Arman-Mojaver) [inline main.py]: same here

File: main.py [hunk 8]
```diff
@@ -1 +1,107 @@
...
+                row.Cost
+            )
+        conn.commit()
+        print('Original Data is loaded.')
+
+    except Exception as e:
+        print(f"Error loading data: {e}")
+
+def load_incremental_data(new_file, cursor, conn):
+    try:
```
Comment (Arman-Mojaver) [inline main.py]: this try block is too large, which many things can fail, and everything is being caught with the general Exception type. Extract what can not fail (or is unlikely to fail) from the try block (for example the pandas stuff

File: main.py [hunk 9]
```diff
@@ -1 +1,107 @@
...
+
+        # If there is no data in the database, set a minium date as 1900-01-01
+        if max_date is None:
+            max_date = pd.to_datetime('1900-01-01')
+
+        # Filter out incremental data with dates greater than the maximum date in the database
+        new_data = incremental_list[incremental_list['Date'] > max_date]
+
+        # If there is new_data, insert the new_data into database
+        if not new_data.empty:
```
Comment (Arman-Mojaver) [inline main.py]: do not use if else statements, use early returns

File: main.py [hunk 10]
```diff
@@ -1 +1,107 @@
...
+        CREATE TABLE masterlist (
+            Date datetime,
+            Server nvarchar(200),
+            Cost float
+            )
+        ''')
+        conn.commit()
+        print('Table is Created!')
+
+    except Exception as e:
```
Comment (Arman-Mojaver) [inline main.py]: same here

File: main.py [hunk 11]
```diff
@@ -1 +1,107 @@
...
+                conn.commit()
+            print('Incremental data is loaded.')
+        else:
+                print('No new data to load.')
+
+    except Exception as e:
+        print(f'Error loading incremental data: {e}')
+
+
+server_name = 'Jiani_da\MSSQLSERVER04'
```
Comment (Arman-Mojaver) [inline main.py]: use upper case for these types of variables that really are constants --> SERVER_NAME = ...

File: main.py [hunk 12]
```diff
@@ -1 +1,107 @@
...
+database_name = 'Python-SQL-CSV'
+original_file = '2023 Jan data.csv'
+new_file = '2023 Mar data.csv'
+# Calling the function to connect to the database
+conn, cursor = database_connect(server_name, database_name)
+# Calling the function to create table in database
+#create_dbtable(original_file, cursor, conn)
```
Comment (Arman-Mojaver) [inline main.py]: delete commented out code

**Review by Arman-Mojaver — CHANGES_REQUESTED** [review]
From a design point of view, it would be great to use one or several classes, instead of several functions.

File: main.py (hunk 1 above)
Comment (Jiani-Li) [reply main.py]: Can you give me an example? Thanks!

Comment (Arman-Mojaver) [reply main.py]: something like this, but with a better print statement (maybe returning the parameters you introduced in the print statement):
https://stackoverflow.com/a/34312912

File: main.py (hunk 9 above)
Comment (Jiani-Li) [reply main.py]: Could you give me an example please?Thanks

**Comment by Jiani-Li** [comment]
> From a design point of view, it would be great to use one or several classes, instead of several functions.

Thanks for your suggestion! I would like to create 2 class: 1. CsvToDatabase 2. DatabaseConnectionError(Exception) In the first class should I use the try and except clause with class of DatabaseConnectionError(Exception) ?

**Comment by Arman-Mojaver** [comment]
> > From a design point of view, it would be great to use one or several classes, instead of several functions.
> 
> Thanks for your suggestion! I would like to create 2 class: 1. CsvToDatabase 2. DatabaseConnectionError(Exception) In the first class should I use the try and except clause with class of DatabaseConnectionError(Exception) ?

Yes, that sounds good. the DatabaseConnectionError(Exception) is only two lines, exactly as I wrote it. The CsvToDatabase  class contains most of your code

**Pull request merged**